pkg/kubelet/dockershim/docker_stats.go

```diff
@@ -17,16 +17,112 @@ limitations under the License.
+	"runtime"
+	"time"
 )
 
-// DockerService does not implement container stats.
-func (ds *dockerService) ContainerStats(*runtimeapi.ContainerStatsRequest) (*runtimeapi.ContainerStatsResponse, error) {
-	return nil, fmt.Errorf("Not implemented")
+// ContainerStats returns stats for a container stats request based on container id.
+func (ds *dockerService) ContainerStats(containerStatsRequest *runtimeapi.ContainerStatsRequest) (*runtimeapi.ContainerStatsResponse, error) {
+	containerStats, err := ds.createContainerStats(containerStatsRequest.ContainerId)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &runtimeapi.ContainerStatsResponse{
+		Stats: containerStats,
+	}, nil
 }
 
-func (ds *dockerService) ListContainerStats(*runtimeapi.ListContainerStatsRequest) (*runtimeapi.ListContainerStatsResponse, error) {
-	return nil, fmt.Errorf("Not implemented")
+// ListContainerStats returns stats for a list container stats request based on a filter.
+func (ds *dockerService) ListContainerStats(containerStatsRequest *runtimeapi.ListContainerStatsRequest) (*runtimeapi.ListContainerStatsResponse, error) {
+	filter := &runtimeapi.ContainerFilter{}
+
+	if containerStatsRequest.Filter != nil {
+		filter.Id = containerStatsRequest.Filter.Id
+		filter.PodSandboxId = containerStatsRequest.Filter.PodSandboxId
+		filter.LabelSelector = containerStatsRequest.Filter.LabelSelector
+	}
+
+	containers, err := ds.ListContainers(filter)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var stats []*runtimeapi.ContainerStats
+	for _, container := range containers {
+		containerStats, err := ds.createContainerStats(container.Id)
+
+		if err != nil {
+			return nil, err
+		}
+
+		stats = append(stats, containerStats)
+	}
+
+	return &runtimeapi.ListContainerStatsResponse{
+		Stats: stats,
+	}, nil
+}
+
+func (ds *dockerService) createContainerStats(containerID string) (*runtimeapi.ContainerStats, error) {
+	statsJSON, err := ds.client.GetContainerStats(containerID)
+
+	if err != nil {
+		return nil, err
+	}
+
+	containerJSON, err := ds.client.InspectContainerWithSize(containerID)
+
+	if err != nil {
+		return nil, err
+	}
+
+	status, err := ds.ContainerStatus(containerID)
+
+	if err != nil {
+		return nil, err
+	}
+
+	dockerStats := statsJSON.Stats
+	containerStats := &runtimeapi.ContainerStats{
+		Attributes: &runtimeapi.ContainerAttributes{
+			Id:          containerID,
+			Metadata:    status.Metadata,
+			Labels:      status.Labels,
+			Annotations: status.Annotations,
+		},
+		Cpu: &runtimeapi.CpuUsage{
+			Timestamp:            time.Now().UnixNano(),
+			UsageCoreNanoSeconds: &runtimeapi.UInt64Value{Value: ds.getCpu(&dockerStats)},
+		},
+		Memory: &runtimeapi.MemoryUsage{
+			Timestamp:       time.Now().UnixNano(),
+			WorkingSetBytes: &runtimeapi.UInt64Value{Value: ds.getMemory(&dockerStats)},
+		},
+		WritableLayer: &runtimeapi.FilesystemUsage{
+			Timestamp: time.Now().UnixNano(),
+			StorageId: &runtimeapi.StorageIdentifier{Uuid: containerJSON.Driver},
```

Comment:
`containerJSON.Driver` is not the correct info we want. This is the UUID of the filesystem, which is used by kubelet to get the filesystem stats from cadvisor. It's not supported on Windows. I'm not sure we should change cadvisor to support something like this on Windows or probably change CRI to return the stats instead of the UUID in this case.

@yujuhong

Reply:
> This is the UUID of the filesystem, which is used by kubelet to get the filesystem stats from cadvisor. It's not supported on Windows. 

Same questions here: does windows support filesystem UUID? or does all fields (e.g. InodesUsed) in stats API make sense to windows?

Reply:
@bobbypage, could you briefly explain how these fields are mapped to concepts in windows? Thanks.

Reply:
as @yguo0905 mentioned, I don't believe Windows has a UUID of the file system, so there is no equivalent for `StorageId`. Additionally, some of the fields under `FilesystemUsage`, like [`InodesUsed`](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/cri/v1alpha1/runtime/api.pb.go#L3003) don't have corresponding fields on Windows. I will update the PR and remove the `StorageId` usage here.